# 07 오류 처리

*해당 글은 클린코드 도서를 정리한 포스팅입니다.*

## 서론
* 상당수 코드 기반은 전적으로 오류 처리 코드에 좌우된다.
* 여기저리 흩어진 오류 처리 코드 때문에 실제 코드가 하는 일을 파악하기가 거의 불가능 하다는 의미이다.
* 오류 처리는 중요하다. *하지만 오류 처리 코드로 인해 프로그램 논리를 이해하기 어려워진다면 깨끗한 코드라 부르기 어렵다.*


## 목차
* 오류 코드보다 예외를 사용하라
* `Try-Catch-Finally` 문부터 작성하라
* 미확인(unchecked) 예외를 사용하라
* 예외 의미를 제공하라
* 호출자를 고려해 예외 클래스를 정의하라
* 정상 흐름을 정의하라
* null을 반환하지 마라
* null을 전달하지 마라


## 오류 코드보다 예외를 사용하라

### 리팩토링 전
```java
public class DeviceController {
    ...
    public void sendShutDown() {
        DeviceHandle handle = getHandle(DEV1);
        // 디바이스 상태를 점검한댜.
        if (handle != DeviceHandle.INVALID) {
            // 레코드 필드에 디바이스 상태를 저장한다.
            retrieveDeviceRecord(handle);
            // 디바이스가 일시정지 상태가 아니라면 종료한다.
            if (record.getStatus() != DEVICE_SUSPENDED) {
                pauseDevice(handle);
                clearDeviceWorkQueue(handle);
                closeDevice(handle);
            } else {
                logger.log("Device suspended. Unable to shut down");
            }
        } else {
            logger.log("Invalid handle for: " + DEV1.toString());
        }
    }
    ...
}
```
### 단점
* 함수를 호출한 즉시 오류를 확인하는 코드는 호출자 코드가 복잡해진다.
* 이런 코드들은 잊어 버리기 쉽다

** 래서 오류가 발생하면 예외를 던지는 편이 낫다. 그러면 호출자 코드가 더 깔끔해진다.**


### 리팩토링 후
```
public class DeviceController {
    ...
    public void sendShutDown() {
        try {
            tryToShutDown();
        } catch (DeviceShutDownError e) {
            logger.log(e);
        }
    }

    private void tryToShutDown() throws DeviceShutDownError {
        DeviceHandle handle = getHandle(DEV1);
        DeviceRecord record = retrieveDeviceRecord(handle);
        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
    }

    private DeviceHandle getHandle(DeviceID id) {
        ...
        throw new DeviceShutDownError("Invalid handle for: " + id.toString());
        ...
    }
    ...
}
```
* 오류 처리 코드와 비지니스로직 코드가 뒤썩이지 않아 호출자 코드가 단순해진다.

## `Try-Catch-Finally` 문부터 작성하라
* `Try-Catch-Finally` 문에서 try 블록에 들어가는 코드를 실행하면 어느 시점에서든 실행이 중단된 후 catch 블록으로 넘어갈 수 있다.
* 그럼으로 예외가 발생할 코드를 짤 때는 `Try-Catch-Finally` 문으로 시작하는 편이 좋다.
* 그러면 `try` 블록에 무슨 일이 생기든지 호출자가 기대하는 상태를 정의하기가 쉬워진다.
